### ISSUE TITLE
`/search`ing by date and/or datetime and/or datetime+ms

### ISSUE THREAD

**Issue by jgaucher-cs** [stac-utils/stac-fastapi#768]
Hi, when calling the /search endpoint that is implemented here: https://github.com/stac-utils/stac-fastapi/blob/main/stac_fastapi/api/stac_fastapi/api/app.py, it seems we cannot search by date, but only by datetime. 

A GET request with `/search?datetime=2000-01-01` will fail with this error from https://github.com/stac-utils/stac-fastapi/blob/main/stac_fastapi/types/stac_fastapi/types/rfc3339.py
```
b'{"detail":"Invalid RFC3339 datetime."}'
```
A POST request will fail with what seems to by a pydantic error: 
```
b'{"detail":[{"type":"datetime_parsing","loc":["body","datetime"],"msg":"Input should be a valid datetime, invalid datetime separator, expected `T`, `t`, `_` or space","input":"2000-01-01","ctx":{"error":"invalid datetime separator, expected `T`, `t`, `_` or space"}}]}'
```
We would like to allow our users to search by date without time, then our application (that uses stac_fastapi) would return all products for this whole day. Would this be possible ?

Then we would like to do the same for milliseconds: our products have a ms information, so if the user searches by datetime+ms, we would like to return products only for this specific ms. If he searches by datetime without ms, we would like to return all products for this whole second.

Our problem is that, if he searches for ms=.000 (to have products only for this specific ms) with a GET request e.g. `/search?datetime=2000-01-01T01:01:01.000Z`, stac_fastpi **removes the ms=.000 information**, so our app doesn't know anymore if the user wanted only products for this specific ms, or for this whole second. I think this removal is made when stac_fastapi converts the str to datetime in `rfc3339_str_to_datetime` https://github.com/stac-utils/stac-fastapi/blob/main/stac_fastapi/types/stac_fastapi/types/rfc3339.py#L51 (I couldn't find where this datetime is converted back to str).

To summarize, my two questions are:

1. Could a user search by date, not datetime ?
2. Could we stop stac_fastpi from removing the "ms=.000" information ?

Thank you :)

**Comment by jgaucher-cs** [comment]
Thank you very much @vincentsarago for your answer. I'm using the pgstac backend.

I will take a look at this customization for searching by date. For the ms, the str to datetime conversion with `rfc3339_str_to_datetime` is working well, the field `microsecond` has the value 0, this is OK. The problem is that datetime is converted back into a str before being passed to my app, and here the ms=.000 is lost.

Allow me to paste a screenshot from my call stack, I've called my endpoint with `/search?datetime=2000-01-01T01:01:01.000Z", then in: /stac_fastapi/pgstac/core.py
![image](https://github.com/user-attachments/assets/58f4db96-755d-460d-800b-bed98c5f769d)

This `p` variable will be passed to my app, we see this is a `list[str]` with one element that contains "datetime":"2000-01-01T01:01:01Z" so here the ms=.000 information is lost. 

A little before in the call stack, we see that `self.post_request_model` returns a datetime as a str (without the ms=.000) that will be used later in the call stack, and also an end_date as a datetime with the ms info, but which is not used later.
![image](https://github.com/user-attachments/assets/7edfd37b-425a-4de1-b31f-18f0115fce01)

**Comment by vincentsarago** [comment]
Can you share the version of stac-fastapi, stac-fastapi-pgstac and stac-pydantic please

**Comment by jgaucher-cs** [comment]
> Can you share the version of stac-fastapi, stac-fastapi-pgstac and stac-pydantic please
I have these: 
```
pypgstac                                 0.8.6                               
pystac                                   1.11.0                              
stac-fastapi-api                         3.0.3                               
stac-fastapi-extensions                  3.0.3                               
stac-fastapi-pgstac                      3.0.0                               
stac-fastapi-types                       3.0.3                               
stac-pydantic                            3.1.3          
```

**Comment by jgaucher-cs** [comment]
Hello @vincentsarago, would you have any update on this ? Thank you :)

**Comment by vincentsarago** [comment]
sorry @jgaucher-cs just got back from vacations, I'll try to have a look this week

**Comment by vincentsarago** [comment]
🤔 in https://github.com/stac-utils/stac-fastapi/pull/771 I've added more tests about `ms` and I don't see anything so I'm pretty sure this is a stac-fastapi-pgstac issue
